### ISSUE TITLE
Support conversion of simdjson_result to dom::element

### ISSUE THREAD

**Issue by mmatrosov** [simdjson/simdjson#1341]
`simdjson_result` is a very transparent structure. It can be implicitly treated as an array with `at()` and `begin()/end()` members, or it can be treated as an object with `operator[]` member. And it can be implicitly converted to `dom::array` and `dom::object`. But not to `dom::element`. I wonder why? I often find myself writing things like `function_acceping_dom_element(json["node"].value())`, note the call to `value()`. I have to do this since there is no implicit conversion. But it seems unnecessary. Are there any reasons why the conversion is not present?

**Comment by lemire** [comment]
Can you give a concrete example?

I expect the following to work...
```C++
void f(dom::element e) {
}

...
f(json["node"]);
```

Of course, as written, it might throw. So it is important to specify whether you are coding with or without exceptions, since we support both.

**Comment by mmatrosov** [comment]
Ok, the problem is a bit trickier.

```
void f(dom::element e) {
}

...
f(json["node"]);  // Works
auto node = json["node"];
f(node); // Does not work
f(std::move(node));  // Works
```

The problem is that conversions to `dom::array` and `dom::object` are defined in `simdjson_result<dom::element>` specialization as 

```
  simdjson_really_inline operator dom::array() const noexcept(false);
  simdjson_really_inline operator dom::object() const noexcept(false);
```

While conversion to `dom::element` is defined in `simdjson_result_base` template as 

```
  simdjson_really_inline operator T&&() && noexcept(false);
```

I.e. it works only on rvalues. And I see no reason why it should not work on lvalues.

**Comment by lemire** [comment]
@jkeiser Got an opinon on this one?

**Comment by jkeiser** [comment]
@mmatrosov that makes sense. I initially wanted to make sure dom::element and friends had to use move semantics, so that we could change the contract for efficiency without breaking people ... but now I think ondemand is going to take on some of the ideas instead. dom::element doesn't have anything uncopiable, so we should allow lvalues to copy!

**Comment by mmatrosov** [comment]
I perceive `dom::element` and friends as "views" which are cheap to copy and pass them by value wherever possible. Seems like it agrees with the examples that I saw. Probably it's time to explicitly document this practice :)

**Comment by lemire** [comment]
@mmatrosov  You are correct. Marked for 0.8 release.

**Comment by lemire** [comment]
@mmatrosov @jkeiser 

We end up with errors due to ambiguity. The compiler does not know whether to move or to copy given the following code sample:

```c++
 dom::element doc = parser.parse("[ true, false ]"_padded);
 doc = parser.parse("[1, 2, 3]"_padded);
```

https://github.com/simdjson/simdjson/pull/1394

**Comment by lemire** [comment]
Revising the motivation.


```C++
void f(dom::element e) {
}

...
f(json["node"]);  // Works
auto node = json["node"];
f(node); // Does not work
f(std::move(node));  // Works

dom::element n = json["node"];
f(n); // works
```

So the issue is trickier that I thought. You want to write a function that takes in `dom::element`, yet you want to call it with a `simdjson_result<dom::element>` value. And you want an automatic translation to occur.

**Comment by lemire** [comment]
The problem is that if `simdjson_result<dom::element>` can go to both dom::element&& and const dom::element&, then it introduces and `doc = parser.parse("[1, 2, 3]"_padded);` can work with both a move and a reference, we have an ambiguity. We could disallow either copying or moving `dom::element`, but which one?

So this issue needs to be better defined.

I will move it off milestone 0.8. Let us revisit later.

**Comment by lemire** [comment]
I am moving it to milestone 2 for a later revisit. We can handle it sooner if someone wants to figure out the right design. Meanwhile, the following works:

```
void f(dom::element e) {
}

dom::element n = json["node"];
f(n); 
```

**Comment by jkeiser** [comment]
Yeah, this is a tough one in general. I think the removal of std::pair might have made it easier, but I think some seriously deep understanding of C++ is enough to do it right.

What I think we want is for `simdjson_result<T>` to be castable to `const T &` when T is copyable. But for move-only classes, we want it to cast to `T&&` if `simdjson_result<T>` is a temporary (`&&`). The issue is, we *don't* want both casts on the same class, or `parser.parse("1"_padded)` becomes ambiguous.

One way to solve this is to have separate simdjson_result_base classes for move-only vs. copy-only classes. Otherwise we might need to use `enable_if`--which might be a great solution as long as it's supported in C++11 across all our compilers!

**Comment by lemire** [comment]
@jkeiser `enable_if` is C++11 so it should be supported everywhere.

It is easy enough to enable just one of the two casts, but one has to specify when one is enabled and not the other. So it is less about C++ (meaning that C++ will allow us to do it for sure if we know what we are doing) and more about spelling out the design in full.
